### ISSUE TITLE
Fix racy header gen cmake depends

### ISSUE THREAD

**Issue by bazaah** [bazaah/aur-ceph#2]
During my work on v17 I found a couple racy compile fails due to missing dependencies on legacy headers the upstream now generates.

- 97e0684826da6cc319e91d3e6670af20e243b35a
- d7e08a7553f614fe9889932fc1ae7101d9dae0f4

Others found more:

- https://github.com/bazaah/aur-ceph/issues/1#issuecomment-1293975296
- https://aur.archlinux.org/pkgbase/ceph#comment-887391

Need to figure out + add the cmake rules to properly depend on those header files existing.

**Comment by petronny** [comment]
```
[  3%] Building C object src/erasure-code/jerasure/CMakeFiles/jerasure_objs.dir/jerasure/src/reed_sol.c.o
cd /build/ceph/src/ceph-17.2.5/build/src/erasure-code/jerasure && /usr/bin/cc -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/build/ceph/src/ceph-17.2.5/build/src/include -I/build/ceph/src/ceph-17.2.5/src -I/build/ceph/src/ceph-17.2.5/src/erasure-code/jerasure -isystem /build/ceph/src/ceph-17.2.5/build/include -isystem /build/ceph/src/ceph-17.2.5/src/xxHash -isystem /build/ceph/src/ceph-17.2.5/src/rapidjson/include -isystem /build/ceph/src/ceph-17.2.5/src/erasure-code/jerasure/jerasure/include -isystem /build/ceph/src/ceph-17.2.5/src/erasure-code/jerasure/gf-complete/include -march=x86-64 -mtune=generic -O2 -pipe  -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection  -DFMT_DEPRECATED_OSTREAM -O2 -g -DNDEBUG -fPIC   -U_FORTIFY_SOURCE -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -fstack-protector-strong -fdiagnostics-color=auto -std=gnu99 -MD -MT src/erasure-code/jerasure/CMakeFiles/jerasure_objs.dir/jerasure/src/reed_sol.c.o -MF CMakeFiles/jerasure_objs.dir/jerasure/src/reed_sol.c.o.d -o CMakeFiles/jerasure_objs.dir/jerasure/src/reed_sol.c.o -c /build/ceph/src/ceph-17.2.5/src/erasure-code/jerasure/jerasure/src/reed_sol.c
In file included from /build/ceph/src/ceph-17.2.5/src/common/config_values.h:59,
                 from /build/ceph/src/ceph-17.2.5/src/common/config.h:28,
                 from /build/ceph/src/ceph-17.2.5/src/common/config_proxy.h:6,
                 from /build/ceph/src/ceph-17.2.5/src/common/ceph_context.h:41,
                 from /build/ceph/src/ceph-17.2.5/src/common/dout.h:29,
                 from /build/ceph/src/ceph-17.2.5/src/include/Context.h:19,
                 from /build/ceph/src/ceph-17.2.5/src/msg/Message.h:24,
                 from /build/ceph/src/ceph-17.2.5/src/msg/DispatchQueue.cc:15:
/build/ceph/src/ceph-17.2.5/src/common/options/legacy_config_opts.h:3:10: fatal error: mds_legacy_options.h: No such file or directory
    3 | #include "mds_legacy_options.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
-- Looking for sched_getcpu - found
-- Looking for getauxval
make[2]: *** [src/msg/CMakeFiles/common-msg-objs.dir/build.make:79: src/msg/CMakeFiles/common-msg-objs.dir/DispatchQueue.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```
Full build log: https://github.com/arch4edu/cactus/actions/runs/3361784541/jobs/5572635705

**Comment by petronny** [comment]
In a new build https://github.com/arch4edu/cactus/actions/runs/3374500877/jobs/5600190817, it failed with:
```
[  2%] Generating rgw_options.cc, ../../../include/rgw_legacy_options.h
cd /build/ceph/src/ceph-17.2.5/build/src/common/options && /usr/bin/python3.10 /build/ceph/src/ceph-17.2.5/src/common/options/y2c.py --input /build/ceph/src/ceph-17.2.5/build/src/common/options/rgw.yaml --output rgw_options.cc --legacy /build/ceph/src/ceph-17.2.5/build/include/rgw_legacy_options.h --name rgw
In file included from /build/ceph/src/ceph-17.2.5/src/common/config_values.h:59,
                 from /build/ceph/src/ceph-17.2.5/src/common/config.h:28,
                 from /build/ceph/src/ceph-17.2.5/src/common/config_proxy.h:6,
                 from /build/ceph/src/ceph-17.2.5/src/common/ceph_context.h:41,
                 from /build/ceph/src/ceph-17.2.5/src/compressor/Compressor.cc:23:
/build/ceph/src/ceph-17.2.5/src/common/options/legacy_config_opts.h:11:10: fatal error: rgw_legacy_options.h: No such file or directory
   11 | #include "rgw_legacy_options.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
But outputs show that `rgw_legacy_options.h` is being generated.

So I think this is an error with paralell compiling, I'm building it with `-j1`.

**Comment by petronny** [comment]
I confirm that both methods fix this error.

**Comment by bazaah** [comment]
> Another way to fix it is `for i in $(seq 5);do VERBOSE=1 make -C build all && break || continue;done`

Yes, it appears d7e08a7553f614fe9889932fc1ae7101d9dae0f4 merely "worked on my system" and didn't actually fix the racy deps.

**Comment by bazaah** [comment]
So I can't figure out how to replicate this. What I'm going to try instead is set up a dummy cmake project that _"generates"_ (read: `cp -v`) a header file with some large delay and experiment with how to force that file to exist before allow cmake to proceed.

**Comment by bazaah** [comment]
Think I have a fix for this: https://github.com/bazaah/aur-ceph/compare/feature/v17.2.5-2

If anyone wants to try it out.

**Comment by bazaah** [comment]
Fixed in 34f0925dea85d9782bfa734119473f6eb708f97a

**Comment by petronny** [comment]
It works. https://github.com/arch4edu/cactus/actions/runs/3402599590/jobs/5667629402
